pythran/cxxgen.py

```diff
@@ -472,64 +476,237 @@ def generate(self):
-        module initialization function.
-        """
-        self.init_body.extend(body)
-
     def add_to_preamble(self, *pa):
         self.preamble.extend(pa)
 
     def add_to_includes(self, *incl):
         self.includes.extend(incl)
 
-    def add_function(self, func, name):
-        """Add a function to be exposed. *func* is expected to be a
+    def add_meta(self, infos):
+        self.infos = infos
+
+    def add_function(self, func, name, types):
+        """
+        Add a function to be exposed. *func* is expected to be a
         :class:`cgen.FunctionBody`.
+
+        Because a function can have several signatures exported,
+        this method actually creates a wrapper for each specialization
+        and a global wrapper that checks the argument types and
+        runs the correct candidate, if any
         """
-        doc = self.docstrings.get(name, '')
-        bpdef = "boost::python::def(\"%s\", &%s, %s)" % (name,
-                                                         func.fdecl.name,
-                                                         self.docstring(doc))
 
-        self.mod_body.append(func)
-        self.init_body.append(Statement(bpdef))
+        self.implems.append(func)
+
+        args_unboxing = []  # turns PyObject to c++ object
+        args_checks = []  # check if the above conversion is valid
+
+        for i, t in enumerate(types):
+            args_unboxing.append('from_python<{}>(args_obj[{}])'.format(t, i))
+            args_checks.append('is_convertible<{}>(args_obj[{}])'.format(t, i))
+        if types:
+            wrapper = '''
+                static PyObject *
+                {ward}wrap_{name}(PyObject *self, PyObject *args)
+                {{
+                    PyObject* args_obj[{size}+1];
+                    PyArg_ParseTuple(args, "{fmt}", {objs});
+                    if({checks})
+                        return to_python({name}({args}));
+                    else {{
+                        return nullptr;
+                    }}
+                }}'''
+        else:
+            wrapper = '''
+                static PyObject *
+                {ward}wrap_{name}(PyObject *self, PyObject *args)
+                {{
+                    return to_python({name}({args}));
+                }}'''
+
+        self.wrappers.append(
+            wrapper.format(name=func.fdecl.name,
+                           size=len(types),
+                           pyname=name,
+                           fmt="O" * len(types),
+                           objs=', '.join('&args_obj[%d]' % i
+                                          for i in range(len(types))),
+                           args=', '.join(args_unboxing),
+                           checks=' and '.join(args_checks),
+                           ward=pythran_ward,
+                           )
+        )
+
+        func_descriptor = pythran_ward + 'wrap_' + func.fdecl.name, types
```

Comment:
If you define this name, you should use it in the previous template to be sure it is the same function name.